File: 089_Number of Islands.py
"""
Number of Islands
Count islands of '1's in a 2D grid using DFS.
"""
from typing import List

def num_islands(grid: List[List[str]]) -> int:
    if not grid: return 0
    m, n = len(grid), len(grid[0])
    def dfs(i,j):
        if i<0 or j<0 or i>=m or j>=n or grid[i][j] != '1': return
        grid[i][j] = '0'
        dfs(i+1,j); dfs(i-1,j); dfs(i,j+1); dfs(i,j-1)
    cnt = 0
    for i in range(m):
        for j in range(n):
            if grid[i][j] == '1':
                cnt += 1
                dfs(i,j)
    return cnt

if __name__ == '__main__':
    g = [['1','1','0','0','0'],['1','1','0','0','0'],['0','0','1','0','0'],['0','0','0','1','1']]
    print('Number of islands:', num_islands(g))

# Time Complexity: O(m*n)
# Space Complexity: O(m*n)

File: 089_Number of Islands.cpp
// Number of Islands
// Count number of islands in a grid of '1' and '0' using DFS/BFS.
#include <bits/stdc++.h>
using namespace std;

int numIslands(vector<vector<char>>& grid) {
    if(grid.empty()) return 0;
    int m=grid.size(), n=grid[0].size();
    int cnt=0;
    function<void(int,int)> dfs = [&](int i,int j){
        if(i<0||j<0||i>=m||j>=n||grid[i][j]!='1') return;
        grid[i][j]='0';
        dfs(i+1,j); dfs(i-1,j); dfs(i,j+1); dfs(i,j-1);
    };
    for(int i=0;i<m;++i) for(int j=0;j<n;++j) if(grid[i][j]=='1'){ ++cnt; dfs(i,j); }
    return cnt;
}

int main(){
    vector<vector<char>> g = {{'1','1','0','0','0'},{'1','1','0','0','0'},{'0','0','1','0','0'},{'0','0','0','1','1'}};
    cout<<"Number of islands: "<<numIslands(g)<<"\n"; // 3
    return 0;
}

// Time Complexity: O(m*n)
// Space Complexity: O(m*n) recursion in worst case
